### ISSUE TITLE
[BUG] Type errors that only only arises during "PostTypeChecker" and sometimes crashes

### ISSUE THREAD

**Issue by shonfeder** [informalsystems/apalache#1015]
The following spec typechecks ok, and can run to find a counterexample:

```tla
------------------- MODULE sum_types -----------------------
EXTENDS Naturals, FiniteSets
\* VARIABLE x

\* These(A, B, C) == Either(A, Either(B, C))

\* @typeAlias: MSG = [int: Set(Int), str: Set(Str)];
\* @typeAlias: MSGS = Set(MSG);
EXTypeAliases2 == TRUE

VARIABLE
  \* @type: Set(MSG);
  msgs,
  \* @type: Bool;
  found3,
  \* @type: Bool;
  foundC

Ints == {1,2,3}
Strs == {"a","b","c"}

\* Example enum declaration
\* APLEnum("MSG") ==
\*   [ Int: Int
\*   , Str: Str
\*   ]

\* Internal
MsgBase == [ Int |-> {}
           , Str |-> {}
           ]

\* Construtors
Int(i) == [MsgBase EXCEPT !.Int = {i}]
Str(s) == [MsgBase EXCEPT !.Str = {s}]

\* @type: MSG => Bool;
IsInt(m) == Cardinality(m.Int) = 1
\* @type: MSG => Bool;
IsStr(m) == Cardinality(m.Str) = 1

\* @type: (MSG, Int => Bool, Str => Bool) => Bool;
CaseMsg(m, i(_), s(_)) ==
  CASE IsInt(m) -> i(CHOOSE x \in m.Int: TRUE)
    [] IsStr(m) -> s(CHOOSE x \in m.Str: TRUE)

\* @type: MSGS;
Messages == [ Int : {Ints}, Str : {} ]
            \union
            [ Int : {} , Str : {Strs}]

Send(m) == msgs' = msgs \union {m} /\ UNCHANGED <<found3, foundC>>
Rm(m) == msgs' = msgs \ {m}

Rcv ==
  LET
    Check3(x) == x = 3   /\ found3' = TRUE /\ UNCHANGED foundC
    CheckC(x) == x = "c" /\ foundC' = TRUE /\ UNCHANGED found3
  IN
    \E m \in msgs:
        /\ CaseMsg(m, Check3, CheckC)
        /\ Rm(m)

Init ==
    /\ msgs = {}
    /\ found3 = FALSE
    /\ foundC = FALSE

Next ==
  \/ \E i \in Ints: Send(Int(i))
  \/ \E s \in Strs: Send(Str(s))
  \/ Rcv

Inv == Cardinality(msgs) = 2 \/ found3 \/ foundC

TypeOK ==
  /\ msgs \subseteq Messages

------------------------------------------------------------
============================================================
```

But if we switch the `Inv` to use `<` instead of `=` or `>`, then we get a type error on "PostTypeChecker":

```tla
Inv == Cardinality(msgs) = 2 \/ found3 \/ foundC
```

```
PASS #12: PostTypeCheckerSnowcat                                  I@01:46:57.565
 > Running Snowcat .::.                                           I@01:46:57.566
[sum_types.tla:70:6-70:32]: No match between operator signature ((Int, Int) => Bool) and arguments Bool and Bool E@01:46:57.642
[unknown location]: Error when computing the type of Next$0000    E@01:46:57.644
```

And if we replace `2` with `3` we get a crash:

```
PASS #13: BoundedChecker                                          I@01:48:01.134
State 0: Checking 1 state invariants                              I@01:48:01.398
Step 0: picking a transition out of 1 transition(s)               I@01:48:01.404
(Please report an issue at: [https://github.com/informalsystems/apalache/issues],at.forsyte.apalache.tla.lir.TypingException: Unexpected operator type OperT1)
Type checker error: Unexpected operator type OperT1               E@01:48:01.421
at.forsyte.apalache.tla.lir.TypingException: Unexpected operator type OperT1
	at at.forsyte.apalache.tla.bmcmt.types.package$CellT$.fromType1(package.scala:180)
	at at.forsyte.apalache.tla.bmcmt.types.package$CellT$.fromType1(package.scala:168)
	at at.forsyte.apalache.tla.bmcmt.types.package$CellT$.fromTypeTag(package.scala:191)
	at at.forsyte.apalache.tla.bmcmt.rules.SetCtorRule.apply(SetCtorRule.scala:29)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.eachExpr$1(SymbStateRewriterImpl.scala:421)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.$anonfun$rewriteSeqUntilDone$1(SymbStateRewriterImpl.scala:427)
	at scala.collection.immutable.List.map(List.scala:297)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteSeqUntilDone(SymbStateRewriterImpl.scala:427)
	at at.forsyte.apalache.tla.bmcmt.rules.RecCtorRule.apply(RecCtorRule.scala:41)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.FunExceptRule.apply(FunExceptRule.scala:37)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.eachExpr$1(SymbStateRewriterImpl.scala:421)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.$anonfun$rewriteSeqUntilDone$1(SymbStateRewriterImpl.scala:427)
	at scala.collection.immutable.List.map(List.scala:293)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteSeqUntilDone(SymbStateRewriterImpl.scala:427)
	at at.forsyte.apalache.tla.bmcmt.rules.SetCtorRule.apply(SetCtorRule.scala:26)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.SetCupRule.apply(SetCupRule.scala:31)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.AssignmentRule.apply(AssignmentRule.scala:39)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.mapArg$1(AndRule.scala:82)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.$anonfun$apply$2(AndRule.scala:86)
	at scala.collection.immutable.List.map(List.scala:293)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.apply(AndRule.scala:86)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInStaticallyNonEmptySet(QuantRule.scala:233)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInSet(QuantRule.scala:207)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.apply(QuantRule.scala:50)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.trex.TransitionExecutorImpl.prepareTransition(TransitionExecutorImpl.scala:98)
	at at.forsyte.apalache.tla.bmcmt.trex.FilteredTransitionExecutor.prepareTransition(FilteredTransitionExecutor.scala:34)
	at at.forsyte.apalache.tla.bmcmt.trex.ConstrainedTransitionExecutor.prepareTransition(ConstrainedTransitionExecutor.scala:93)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.$anonfun$prepareTransitionsAndCheckInvariants$6(SeqModelChecker.scala:147)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.$anonfun$prepareTransitionsAndCheckInvariants$6$adapted(SeqModelChecker.scala:141)
	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:985)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:984)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.prepareTransitionsAndCheckInvariants(SeqModelChecker.scala:141)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.makeStep(SeqModelChecker.scala:63)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.run(SeqModelChecker.scala:51)
	at at.forsyte.apalache.tla.bmcmt.passes.BoundedCheckerPassImpl.runIncrementalChecker(BoundedCheckerPassImpl.scala:131)
	at at.forsyte.apalache.tla.bmcmt.passes.BoundedCheckerPassImpl.execute(BoundedCheckerPassImpl.scala:98)
	at at.forsyte.apalache.infra.passes.PassChainExecutor.exec$1(PassChainExecutor.scala:22)
	at at.forsyte.apalache.infra.passes.PassChainExecutor.run(PassChainExecutor.scala:37)
	at at.forsyte.apalache.tla.Tool$.runCheck(Tool.scala:187)
	at at.forsyte.apalache.tla.Tool$.$anonfun$run$3(Tool.scala:95)
	at at.forsyte.apalache.tla.Tool$.handleExceptions(Tool.scala:322)
	at at.forsyte.apalache.tla.Tool$.run(Tool.scala:95)
	at at.forsyte.apalache.tla.Tool$.main(Tool.scala:45)
	at at.forsyte.apalache.tla.Tool.main(Tool.scala)
```

**Comment by shonfeder** [comment]
If I move the `Check` operators into the top level I get a different crash: 

```
State 0: Checking 1 state invariants                              I@01:49:29.482
Step 0: picking a transition out of 1 transition(s)               I@01:49:29.488
State 1: Checking 1 state invariants                              I@01:49:29.524
(Please report an issue at: [https://github.com/informalsystems/apalache/issues],at.forsyte.apalache.tla.lir.TypingException: Unexpected type VarT1)
Type checker error: Unexpected type VarT1                         E@01:49:29.533
at.forsyte.apalache.tla.lir.TypingException: Unexpected type VarT1
	at at.forsyte.apalache.tla.bmcmt.types.package$CellT$.fromType1(package.scala:166)
	at at.forsyte.apalache.tla.bmcmt.types.package$CellT$.fromType1(package.scala:168)
	at at.forsyte.apalache.tla.bmcmt.types.package$CellT$.fromTypeTag(package.scala:191)
	at at.forsyte.apalache.tla.bmcmt.rules.SetCtorRule.apply(SetCtorRule.scala:29)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.eachExpr$1(SymbStateRewriterImpl.scala:421)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.$anonfun$rewriteSeqUntilDone$1(SymbStateRewriterImpl.scala:427)
	at scala.collection.immutable.List.map(List.scala:293)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteSeqUntilDone(SymbStateRewriterImpl.scala:427)
	at at.forsyte.apalache.tla.bmcmt.rules.RecCtorRule.apply(RecCtorRule.scala:41)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.FunExceptRule.apply(FunExceptRule.scala:37)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.eachExpr$1(SymbStateRewriterImpl.scala:421)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.$anonfun$rewriteSeqUntilDone$1(SymbStateRewriterImpl.scala:427)
	at scala.collection.immutable.List.map(List.scala:293)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteSeqUntilDone(SymbStateRewriterImpl.scala:427)
	at at.forsyte.apalache.tla.bmcmt.rules.SetCtorRule.apply(SetCtorRule.scala:26)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.SetCupRule.apply(SetCupRule.scala:31)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.AssignmentRule.apply(AssignmentRule.scala:39)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.mapArg$1(AndRule.scala:82)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.$anonfun$apply$2(AndRule.scala:86)
	at scala.collection.immutable.List.map(List.scala:293)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.apply(AndRule.scala:86)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInStaticallyNonEmptySet(QuantRule.scala:233)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.skolemExistsInSet(QuantRule.scala:207)
	at at.forsyte.apalache.tla.bmcmt.rules.QuantRule.apply(QuantRule.scala:50)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:336)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive$1(SymbStateRewriterImpl.scala:367)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:401)
	at at.forsyte.apalache.tla.bmcmt.trex.TransitionExecutorImpl.prepareTransition(TransitionExecutorImpl.scala:98)
	at at.forsyte.apalache.tla.bmcmt.trex.FilteredTransitionExecutor.prepareTransition(FilteredTransitionExecutor.scala:34)
	at at.forsyte.apalache.tla.bmcmt.trex.ConstrainedTransitionExecutor.prepareTransition(ConstrainedTransitionExecutor.scala:93)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.$anonfun$prepareTransitionsAndCheckInvariants$6(SeqModelChecker.scala:147)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.$anonfun$prepareTransitionsAndCheckInvariants$6$adapted(SeqModelChecker.scala:141)
	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:985)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:984)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.prepareTransitionsAndCheckInvariants(SeqModelChecker.scala:141)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.makeStep(SeqModelChecker.scala:63)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.run(SeqModelChecker.scala:51)
	at at.forsyte.apalache.tla.bmcmt.passes.BoundedCheckerPassImpl.runIncrementalChecker(BoundedCheckerPassImpl.scala:131)
	at at.forsyte.apalache.tla.bmcmt.passes.BoundedCheckerPassImpl.execute(BoundedCheckerPassImpl.scala:98)
	at at.forsyte.apalache.infra.passes.PassChainExecutor.exec$1(PassChainExecutor.scala:22)
	at at.forsyte.apalache.infra.passes.PassChainExecutor.run(PassChainExecutor.scala:37)
	at at.forsyte.apalache.tla.Tool$.runCheck(Tool.scala:187)
	at at.forsyte.apalache.tla.Tool$.$anonfun$run$3(Tool.scala:95)
	at at.forsyte.apalache.tla.Tool$.handleExceptions(Tool.scala:322)
	at at.forsyte.apalache.tla.Tool$.run(Tool.scala:95)
	at at.forsyte.apalache.tla.Tool$.main(Tool.scala:45)
	at at.forsyte.apalache.tla.Tool.main(Tool.scala)
```

**Comment by shonfeder** [comment]
This is just about the length of the run. Replacing the `Cardinality` check with `TRUE` also yields the OperT1 crash:

```
Inv == TRUE \/ found3 \/ foundC
```

**Comment by shonfeder** [comment]
The problems seems to be in the next, where I'm composing operators. If I lift the operator applications using `LET` like so:

```tla
Next ==
  \/ \E i \in Ints: LET x == Int(i) IN Send(x)
  \/ \E s \in Strs: LET x == Str(s) IN Send(x)
  \/ Rcv
```

Then I get the following error:

```
[unknown location]: Operator x$1 has a parameterized type, while polymorphism is disabled: (() => [Int: Set(Int), Str: Set(a367)]) E@01:57:10.754
```

**Comment by shonfeder** [comment]
lol: when I lower-cased the record field names it fixed the problem.

Assuming that's the root cause, I guess we should either raise an error with upper case record fields names or else fix this support them?

**Comment by shonfeder** [comment]
Ok, looks also like operator applications are not being accounted for correctly in the "PostTypeChecker". If I rewrite `Next` like this (after fixing the record case problem), I get an incorrectly typed operator:

```tla
Next ==
  \/ \E i \in Ints: Send(Int(i))
  \/ \E s \in Strs: Send(Str(s))
  \/ Rcv
```

```
[sum_types.tla:69:6-69:32]: No match between operator signature (([int: Set(Int), str: Set(Str)]) => [int: Set(Int), str: Set(Str)]) and argument [int: Set(Int), str: Set(((Int) => Int))] E@02:09:19.389
```

**Comment by konnov** [comment]
This looks like a yet another instance of an incorrect record field access. So we should finally fix #401.

**Comment by konnov** [comment]
Indeed, this issues was caused by imprecise type checking for records. I have changed the type to `{...}`:

```tla
------------------- MODULE sum_types -----------------------
EXTENDS Naturals, FiniteSets
\* VARIABLE x

\* These(A, B, C) == Either(A, Either(B, C))

\* @typeAlias: MSG = {int: Set(Int), str: Set(Str)};
\* @typeAlias: MSGS = Set(MSG);
EXTypeAliases2 == TRUE

VARIABLE
  \* @type: Set(MSG);
  msgs,
  \* @type: Bool;
  found3,
  \* @type: Bool;
  foundC

Ints == {1,2,3}
Strs == {"a","b","c"}

\* Example enum declaration
\* APLEnum("MSG") ==
\*   [ Int: Int
\*   , Str: Str
\*   ]

\* Internal
MsgBase == [ Int |-> {}
           , Str |-> {}
           ]

\* Construtors
Int(i) == [MsgBase EXCEPT !.Int = {i}]
Str(s) == [MsgBase EXCEPT !.Str = {s}]

\* @type: MSG => Bool;
IsInt(m) == Cardinality(m.Int) = 1
\* @type: MSG => Bool;
IsStr(m) == Cardinality(m.Str) = 1

\* @type: (MSG, Int => Bool, Str => Bool) => Bool;
CaseMsg(m, i(_), s(_)) ==
  CASE IsInt(m) -> i(CHOOSE x \in m.Int: TRUE)
    [] IsStr(m) -> s(CHOOSE x \in m.Str: TRUE)

\* @type: MSGS;
Messages == [ Int : {Ints}, Str : {} ]
            \union
            [ Int : {} , Str : {Strs}]

Send(m) == msgs' = msgs \union {m} /\ UNCHANGED <<found3, foundC>>
Rm(m) == msgs' = msgs \ {m}

Rcv ==
  LET
    Check3(x) == x = 3   /\ found3' = TRUE /\ UNCHANGED foundC
    CheckC(x) == x = "c" /\ foundC' = TRUE /\ UNCHANGED found3
  IN
    \E m \in msgs:
        /\ CaseMsg(m, Check3, CheckC)
        /\ Rm(m)

Init ==
    /\ msgs = {}
    /\ found3 = FALSE
    /\ foundC = FALSE

Next ==
  \/ \E i \in Ints: Send(Int(i))
  \/ \E s \in Strs: Send(Str(s))
  \/ Rcv

Inv == Cardinality(msgs) = 2 \/ found3 \/ foundC

TypeOK ==
  /\ msgs \subseteq Messages

------------------------------------------------------------
============================================================
```

Now, the typechecker finds an incorrect record access:

```sh
apalache-mc check --features=rows sum_types.tla
...
PASS #1: TypeCheckerSnowcat                                       I@15:13:17.727
 > Running Snowcat .::.                                           I@15:13:17.727
[sum_types.tla:38:25-38:29]: Cannot apply m to the argument "Int" in m["Int"]. E@15:13:17.896
[sum_types.tla:38:1-38:34]: Error when computing the type of IsInt E@15:13:17.899
```

So I am closing this issue, as #401 has fixed it.
